### PR TITLE
Numeric changes

### PR DESCRIPTION
**Pull request by camembertelectrique**
### What does this PR do?
Speed up byte array to hex string conversion

### Where should the reviewer start?
Method toHexString() reimplemented

### Why is it needed?
Two orders of magnitude faster implementation

### CODE REVIEW

**Comment by conor10** [comment]
Thanks for taking the time to submit this.

Whilst I appreciate it is a speedup, is this sort of optimisation really necessary at present in Web3j given the main constraint we have is the EVM? 

It would be helpful to understand what was the underlying motivation that led you to this specific change, as it sacrifices readability for performance.

**Comment by camembertelectrique** [comment]
Conor, we have an app that generates non-EVM-based raw transactions to a precompiled contract that have large payloads compared to an EVM transaction, and we use Numeric.toHexString() to encode them. During testing, we found that this call created a significant slowdown in our app. This patch resulted in a two orders of magnitude reduction in compute time, and solved the performance issue we were having.

File: utils/src/main/java/org/web3j/utils/Numeric.java [hunk 1]
```diff
@@ -227,15 +228,18 @@ private static String toHexStringZeroPadded(BigInteger value, int size, boolean
...
-        }
-        for (int i = offset; i < offset + length; i++) {
-            stringBuilder.append(String.format("%02x", input[i] & 0xFF));
-        }
+        final String output = new String(toHexCharArray(input, offset, length, withPrefix));
+        return withPrefix ? new StringBuilder(HEX_PREFIX).append(output).toString() : output;
+    }
 
-        return stringBuilder.toString();
+    public static char[] toHexCharArray(byte[] input, int offset, int length, boolean withPrefix) {
```
Comment (conor10) [inline utils/src/main/java/org/web3j/utils/Numeric.java:235]: why is `withPrefix` still here?

Comment (camembertelectrique) [reply utils/src/main/java/org/web3j/utils/Numeric.java:235]: It's to preserve the functionality of toHexString and to avoid an interface change.

Comment (conor10) [reply utils/src/main/java/org/web3j/utils/Numeric.java:235]: but `toHexCharArray` is only defined in this class and could be private

Comment (camembertelectrique) [reply utils/src/main/java/org/web3j/utils/Numeric.java:235]: Yes, you're right. `toHexCharArray` could be made private to preserve the interface as it is.

Comment (conor10) [reply utils/src/main/java/org/web3j/utils/Numeric.java:235]: If you make this update I'm happy to merge.

**Comment by esaulpaugh** [comment]
you can also eliminate one array access per byte with a size-256 encoding table:

```java
final int[] ints = new int[] { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f' };
final int leftNibbleMask = 0xF0;
final int rightNibbleMask = 0x0F;
for (int i = 0; i < ENCODE_TABLE.length; i++) {
    int leftChar = ints[(i & leftNibbleMask) >>> Byte.SIZE / 2];
    int rightChar = ints[i & rightNibbleMask];
    ENCODE_TABLE[i] = (leftChar << Byte.SIZE) | rightChar;
}
```

```java
int hexPair = ENCODE_TABLE[input[offset] & 0xFF];
output[j] = (char) (hexPair >>> Byte.SIZE); // left
output[j+1] = (char) (hexPair & 0xFF); // right
```

resulting in a decent speedup https://github.com/esaulpaugh/headlong/blob/master/src/main/java/com/esaulpaugh/headlong/util/FastHex.java

And if you're optimizing for Java 9+ it's faster to encode to bytes and then use `new String(byte[],int,int,int)` because of CompactStrings.

File: utils/src/main/java/org/web3j/utils/Numeric.java (hunk 1 above)
Comment (camembertelectrique) [reply utils/src/main/java/org/web3j/utils/Numeric.java]: I reviewed and tested the suggestion for performance improvement from @esaulpaugh but found none. I have nothing else to add to this pull, it's good to merge if you approve

Comment (conor10) [reply utils/src/main/java/org/web3j/utils/Numeric.java]: what about the removal of `withPrefix`?

**Pull request merged**